Mathlib/Lean/Meta/RefinedDiscrTree/Encode.lean

```diff
@@ -183,326 +74,372 @@ where
+      mkNewStar
+
+  | .forallE .. =>
+    setEAsPrevious
+    withLams lambdas .forall
+  | .lit v      => withLams lambdas <| .lit v
+  | .sort _     => withLams lambdas <| .sort
+  | .letE ..    => withLams lambdas <| .opaque
+  | .lam ..     => withLams lambdas <| .opaque
   | _           => unreachable!
 
-private abbrev M := StateListT (AssocList Expr DTExpr) <| ReaderT Context MetaM
-
-/-
-Caching values is a bit dangerous, because when two expressions are be equal and they live under
-a different number of binders, then the resulting De Bruijn indices are offset.
-In practice, getting a `.bvar` in a `DTExpr` is very rare, so we exclude such values from the cache.
--/
-instance : MonadCache Expr DTExpr M where
-  findCached? e := do
-    let s ← get
-    return s.find? e
-  cache e e' :=
-    if e'.hasLooseBVars then
-      return
-    else
-      modify (·.insert e e')
+/-- Check whether the expression is represented by `Key.star` and has `arg` as an argument. -/
+private def isStarWithArg (arg : Expr) : Expr → Bool
+  | .app f a => if a == arg then f.getAppFn.isMVar else isStarWithArg arg f
+  | _ => false
 
-/-- Return all pairs of body, bound variables that could possibly appear due to η-reduction -/
-@[specialize]
-def etaPossibilities (e : Expr) (lambdas : List FVarId) (k : Expr → List FVarId → M α) : M α :=
-  k e lambdas
-  <|> do
-  match e, lambdas with
+/-- Run `k` on all pairs of body, bound variables that could possibly appear due to η-reduction -/
+private def etaPossibilities (e : Expr) (lambdas : List FVarId) (root : Bool) (entry : LazyEntry) :
+    ReaderT Context MetaM (List (Key × LazyEntry)) :=
+  return (← (encodingStepAux e lambdas root).run (← read) entry) :: (← match e, lambdas with
   | .app f a, fvarId :: lambdas =>
-    if isStarWithArg (.fvar fvarId) a then
-      withReader (fun c => { c with forbiddenVars := fvarId :: c.forbiddenVars }) do
-        etaPossibilities f lambdas k
+    if isStarWithArg (.fvar fvarId) a && !f.getAppFn.isMVar then
+      etaPossibilities f lambdas root entry
     else
-      failure
-  | _, _ => failure
+      return []
+  | _, _ => return [])
 
 /-- run `etaPossibilities`, and cache the result if there are multiple possibilities. -/
-@[specialize]
-def cacheEtaPossibilities (e original : Expr) (lambdas : List FVarId)
-    (k : Expr → List FVarId → M DTExpr) : M DTExpr :=
+private def cacheEtaPossibilities (e original : Expr) (lambdas : List FVarId) (root : Bool)
+    (entry : LazyEntry) : ReaderT Context MetaM (List (Key × LazyEntry)) := do
   match e, lambdas with
-  | .app _ a, fvarId :: _ =>
-    if isStarWithArg (.fvar fvarId) a then
-      checkCache original fun _ =>
-        etaPossibilities e lambdas k
+  | .app f a, fvarId :: _ =>
+    if isStarWithArg (.fvar fvarId) a && !f.getAppFn.isMVar then
+      match entry.cache.find? original with
+      | some (key :: keys) => return [(key, { entry with results := keys })]
+      | some [] => unreachable!
```

Comment:
Could you add a comment explaining why this is unreachable? Or perhaps `panic! "the reason"` if you're not actually certain?